webapp/api/github.py

```diff
@@ -0,0 +1,139 @@
+import flask
+from urllib.parse import urlencode
+from webapp import api
+from requests.exceptions import RequestException
+
+
+class GithubAPIError(RequestException):
+    pass
+
+
+class GithubAPI:
+    """
+    Provides authentication for GitHub users. Helper methods are also provided
+    for checking organization access and getting user data from the Github API.
+    """
+
+    BASE_URL = "https://api.github.com/"
+    AUTH_URL = "https://github.com/login/oauth/"
+
+    def __init__(
+        self, client_id, client_secret, session=api.requests.Session()
+    ):
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.session = session
+        self.session.headers["Accept"] = "application/json"
+
+    def authorize(self, scope=None, state=None):
+        """
+        Redirect to Github and request access to a user's data.
+        """
+        params = {
+            "client_id": self.client_id,
+        }
+
+        if scope:
+            params["scope"] = scope
+
+        if state:
+            params["state"] = state
+
+        auth_url = f"{self.AUTH_URL}authorize?{urlencode(params)}"
+
+        return flask.redirect(auth_url)
+
+    def verify_auth(self, code):
+        """
+        Handles response after the redirect to Github. This response
+        determines if the user has allowed this application access. If we
+        were then we send a POST request for the access_key used to
+        authenticate requests to Github.
+        """
+        data = {
+            "code": code,
+            "client_id": self.client_id,
+            "client_secret": self.client_secret,
+        }
+
+        response = self.request("POST", self.AUTH_URL, "access_token", data)
+
+        response.raise_for_status()
+
+        data = response.json()
+
+        if "error" in data:
+            raise GithubAPIError(data["error_description"], response=response)
+
+        return data["access_token"]
+
+    def request(
+        self, method, base_url, resource, data={}, params={}, access_token=None
+    ):
+        """
+        Makes a raw HTTP request and returns the response.
+        """
+        url = base_url + resource
```

Comment:
It's a little weird that you're asking for this as 2 separate arguments and then simply concatenating them. Why don't you just ask for one `url`?

Reply:
Although if we do make this class more single-purpose about only making direct authenticated API calls, then maybe you can remove the `base_url` argument, and make this line:

```suggestion
        url = os.path.join(self.BASE_URL, resource)
```